driver/chain_syncer/calldata/syncer.go

```diff
@@ -517,65 +521,123 @@ func (s *Syncer) createExecutionPayloads(
 
 // checkLastVerifiedBlockMismatch checks if there is a mismatch between protocol's last verified block hash and
 // the corresponding L2 EE block hash.
-func (s *Syncer) checkLastVerifiedBlockMismatch(ctx context.Context) (bool, error) {
+func (s *Syncer) checkLastVerifiedBlockMismatch(ctx context.Context) (*rpc.ReorgCheckResult, error) {
+	var (
+		reorgCheckResult = new(rpc.ReorgCheckResult)
+		err              error
+	)
+
 	stateVars, err := s.rpc.GetProtocolStateVariables(&bind.CallOpts{Context: ctx})
 	if err != nil {
-		return false, err
+		return reorgCheckResult, err
 	}
 
 	if s.state.GetL2Head().Number.Uint64() < stateVars.B.LastVerifiedBlockId {
-		return false, nil
+		return reorgCheckResult, nil
 	}
 
-	blockInfo, err := s.rpc.GetL2BlockInfo(ctx, new(big.Int).SetUint64(stateVars.B.LastVerifiedBlockId))
+	reorgCheckResult, err = s.retrievePastBlock(ctx, stateVars.B.LastVerifiedBlockId, 0)
 	if err != nil {
-		return false, err
+		return reorgCheckResult, err
+	}
+
+	return reorgCheckResult, nil
+}
+
+// retrievePastBlock find proper L1 header and L2 block id to reset when there is a mismatch
+func (s *Syncer) retrievePastBlock(ctx context.Context, blockID uint64, retries uint64) (*rpc.ReorgCheckResult, error) {
```

Comment:
Shall we add some unit tests for this new method?

Reply:
Yes, i will add them